### ISSUE TITLE
proposal: function currying

### ISSUE THREAD

**Issue by kyle-github** [zig-lang/zig#491]
This is fallout from the discussion in #484.

From that discussion, it seems that Zig is very close to supporting function currying at compile time.  At least that is what I gathered from the comments.  The exact syntax is not part of this proposal though the cleanest one I saw in #484 was this (from @raulgrell and @andrewrk):

```
fn addOne(n: u8) -> u8 { return add(n, 1); }
```

I am putting this here to separate it out from the original proposal in #484 for default parameters.

Whether function currying is something that is needed in Zig is a different story, but I could see a lot of interesting applications with fake OO and some other interesting things with shared counters (wrap a function around a shared counter) etc.   Currying has a lot of uses (and abuses).

**Comment by thejoshwolfe** [comment]
This issue is not yet a proposal. Care to elaborate on what currying is and how it relates to Zig? Wikipedia's description seems pretty abstract compared to Zig's design goals.

**Comment by kyle-github** [comment]
Currying is a way of creating a function from another function.   During the process, you bind one or more arguments to the original function.  In the examples in #484, you can see one of the examples which I reproduced above.   

Currying is mostly used in functional programming and you can consider it a limited form of continuations.

However, I've used something like this to good effect in languages like Perl where you create a function that encapsulates some other data and then pass that function on without having to pass the data as well.  

Whether that is something that could be done meaningfully within Zig's current framework is a different story.   I wanted to get it on the radar and out of #484 which was diverging far from the original proposal.

I changed the title of the issue to "Discussion" to make it clear that this is not a formal proposal.

**Comment by hasenj** [comment]
I think the terminology you're looking for is partial function application.

A curried function is a bit more complicated.

A partial function application might look like this:

    new_fn = partial_fn(original_fn, arg1=v1, arg4=v4);

where `new_fn` is just a regular function now, it can take some arguments and then pass them along to `original_fn` along with `arg1=v1` and `arg4=v4`.

A curried function is a bit different, I believe. Probably more like this:

   curried_fn = curry_fn(original_fn)

Now `curried_fn` can take an argument and return another curried function, which itself can take an argument and return another curried function, and so on, until all arguments are filled, and it will then call the original function.

so that:

    var f1 = curriend_fn(arg1); // f1 is also a curriend_fn
    var f2 = f1(arg2);
    var f3 = f2(arg3);
    var f4 = f3(arg4);
    ... // after n iterations
    var result = f_n();

**Comment by raulgrell** [comment]
I agree with @hasenj that this is closer to partial function application than currying. Currying implies being able to call any function as if it took only one parameter - the resulting function having one fewer parameter. Consider mem.copy and a curry operator:

```
fn copy(comptime T: type, dest: []T, src: []T) -> []T.

const copybytes = curry copy(u8); // @typeOf(copybytes) == fn([]u8, []u8)->[]u8
const copybytestodest = curry copybytes(slice); // @typeOf(copybytestodest) == fn([]u8)->[]u8)

// If we allowed implicit currying, we'd end up with something like
const copysrcbytestodest = copy(u8)(slice)(src);
```

As it stands the syntax you proposed is valid Zig, and behaves exactly like the above, with the advantage of explicitly specifying parameters. You can somewhat control what really goes on behind the scenes by annotating:

fn addOne(n: u8) -> u8 { return add(n, 1); } - **Function forwarding:** When you call addOne(n), it will then call add(n, 1)

inline fn addOne(n: u8) -> u8 { return add(n, 1); } - **Inline function:** Wherever you call addOne(n), the compiler will replace that function call with add(n, 1).

fn addOne(n: u8) -> u8 { return @inlineCall(add, n, 1); } - **Inline function call:** The body of addOne(n) is identical to the body of the function add, but with a comptime second parameter

**Comment by raulgrell** [comment]
This just so happened to be on the frontpage of HN today:

http://www.datchley.name/currying-vs-partial-application/

**Comment by hasenj** [comment]
Can `partial_fn` be implemented in userspace using some of the comptime features?

**Comment by raulgrell** [comment]
I've actually tried, but it just made me lose some hair. There are a couple of existing proposals that would probably help expressing it like `const new_fn = partial_fn(original_fn, arg1=v1, arg4=v4);`:

- More compile-time reflection - #383
- Better comptime varargs support - #313 and #279

**Comment by hasenj** [comment]
I think what is really needed here to support this in comptime is a way to iteratively build up the parameter list as a heterogeneous tuple before passing it to the function. Which should be possible in theory since a function call in assembly is exactly that.

**Comment by thejoshwolfe** [comment]
Doesn't this require #229? Or are we talking about partial application where the supplied parameters are comptime known?

**Comment by hasenj** [comment]
I think generally speaking, default values are supposed to be known at compile time.

**Comment by thejoshwolfe** [comment]
If this is all at comptime, then what's the difference between partial application and function forwarding? What is this discussion really about?

**Comment by kyle-github** [comment]
Thanks for the comments.   Yes, partial function application/evaluation was closer to where I was aiming.   

I particularly like the different methods of forwarding, and the use of @inline to control how things are built that @raulgrell  showed!

Unless you want environment capture (which is one of the issues, can't remember which one) it seems that Zig already supports most of what I was thinking of.

**Comment by thejoshwolfe** [comment]
> environment capture

closures: #229

**Comment by c-hild** [comment]
> I think the terminology you're looking for is partial function application.

Thank you for clearing this up, also I think the term you're looking for is term :wink: .
